// memshared/lib/Store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
;
var messageId = 0;
var Store = (function () {
    function Store() {
        this.$callbacks = {};
    }
    Store.prototype.dispatch = function (cmd, callback) {
        var args = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            args[_i - 2] = arguments[_i];
        }
        var msg = this._buildMessage.apply(this, [cmd].concat(args));
        // callback when the worker receives back the final result
        if (callback) {
            this.$callbacks[msg.messageId] = callback;
        }
        // send command to be executed by the master node
        process.send(msg);
    };
    Store.prototype.consume = function (message) {
        if (this.$callbacks[message.messageId]) {
            // dispatch callback
            this.$callbacks[message.messageId](message.error, message.result);
            // cleanup
            delete this.$callbacks[message.messageId];
        }
    };
    Store.prototype._buildMessage = function (command) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        return {
            messageId: process.pid + ":" + messageId++,
            cmd: command,
            args: args
        };
    };
    return Store;
}());
exports.Store = Store;
